packages/page-practice/lib/practice/Controller.tsx

```diff
@@ -49,7 +50,7 @@ export const Controller = memo(function Controller({
 function usePracticeState(state: PracticeState) {
   const keyboard = useKeyboard();
   const [lines, setLines] = useState(state.lines); // Forces ui update.
-
+  const { settings } = useSettings();
```

Comment:
A big problem here. You use the `settings` variable` inside a hook, but this variable is not marked as a hook dependency. You should catch this error if you run linter `npm run lint`.